File: scrapper/page.py
import requests
import re
from typing import Optional
from bs4 import BeautifulSoup
from urllib.parse import unquote

CACHE_DIRECTORY = '.cache'
CHARACTER_URL_PATTERN = r"^https:\/\/warriors\.fandom\.com\/wiki\/([^\/:]+)$"

def load_cached(name: str) -> Optional[str]:
    path = f"{CACHE_DIRECTORY}/{name}.html"

    try:
        with open(path, 'r') as file:
            return file.read()

    except FileNotFoundError:
        return None

    except Exception as e:
        printf(f"An exception occured whilst reading file: {path}")
        return None

def store_cached(name: str, content: str):
    path = f"{CACHE_DIRECTORY}/{name}.html"

    try:
        with open(path, 'w') as file:
            file.write(content)

    except Exception as e:
        print(f"An exeption occured whilst storing file: {path}")

def get_page_name(url: str) -> Optional[str]:
    pattern_match = re.match(CHARACTER_URL_PATTERN, url)

    if pattern_match:
        return unquote(pattern_match.group(1)) # Removes HTML encoding

    return None

class Page:
    def __init__(self, url: str):
        self.url = url

    def configure(self) -> bool:
        self.page_name = get_page_name(self.url)

        # The page name is validated, if it doesn't match the string it's
        # straight up just not a character.
        if not self.page_name:
            print(f"Invalid character page url for: {self.url}")
            return False
        
        # The cached version is loaded when possible
        cached_page = load_cached(self.page_name)
        if cached_page:
            self.contents = cached_page
            return True
        
        # If all fails, the page is downloaded directly from Fandom
        response = requests.get(self.url)
        if response.status_code == 200:
            self.contents = response.text
            store_cached(self.page_name, self.contents)
            return True

        else:
            print(f"Request for {self.url} failed with status code: {response.status_code}")
            return False

    def is_character(self) -> bool:
        soup = BeautifulSoup(self.contents, "html.parser")

        for a in soup.select("div.page-header__categories a"):
            if a.get("href") == "/wiki/Category:Characters":
                return True

        return False

    def get_name(self) -> str:
        return self.page_name

    def get_connections(self) -> list[str]:
        if not self.contents:
            return []

        soup = BeautifulSoup(self.contents, "html.parser")

        urls = set()

        for a in soup.find_all("a", href=True):
            url = a["href"]

            # If the page matches the typical
            # page style it can be considered a character
            if get_page_name(url):
                urls.add(url)

        return list(urls)

File: scrapper/page.go
package main

import (
	"errors"
	"fmt"
	"html"
	"io"
	"io/fs"
	"net/http"
	"os"
	"regexp"
	"strings"

	"github.com/PuerkitoBio/goquery"
)

const CacheDirectory = ".cache"

var RegexCharacterURL = regexp.MustCompile(`https:\/\/warriors\.fandom\.com\/wiki\/([^\/:\?#]+)$`)
var RegexCharacterHREF = regexp.MustCompile(`^\/wiki\/([^\/:]+)$`)

var ErrInvalidCharacterUrl = errors.New("invalid character URL")
var ErrPageIsNotCharacter = errors.New("page given is not a character")

func loadCached(name string) (string, error) {
	path := fmt.Sprintf("%s/%s.html", CacheDirectory, name)
	contents, err := os.ReadFile(path)
	return string(contents), err
}

func storeCached(name, content string) error {
	path := fmt.Sprintf("%s/%s.html", CacheDirectory, name)
	return os.WriteFile(path, []byte(content), 0644)
}

func downloadPage(url string) (string, error) {
	resp, err := http.Get(url)
	if err != nil {
		return "", fmt.Errorf("failed to get page %q: %w", url, err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", fmt.Errorf("failed to read page's body %q: %w", url, err)
	}

	return string(body), nil
}

func GetPageName(url string) (string, error) {
	match := RegexCharacterURL.FindStringSubmatch(url)
	if len(match) < 1 {
		return "", ErrInvalidCharacterUrl
	}

	return html.UnescapeString(match[1]), nil
}

func CheckValidUrl(url string) bool {
	_, err := GetPageName(url)
	return err == nil
}

type Page struct {
	URL         string
	Name        string
	IsCharacter bool
	Connections []string
	Contents    string
}

func LoadPage(url string) (*Page, error) {
	pageName, err := GetPageName(url)
	if err != nil {
		return nil, err
	}

	cachedContents, err := loadCached(pageName)
	if err != nil {
		if !errors.Is(err, fs.ErrNotExist) {
			fmt.Printf("Failed to load cache for character %s: %v\n", pageName, err)
			return nil, err
		}

		contents, err := downloadPage(url)
		if err != nil {
			return nil, err
		}

		if err := storeCached(pageName, contents); err != nil {
			fmt.Printf("failed to page %q to cache: %v\n", pageName, err)
		}

		return LoadPageFromContents(url, contents)
	}

	return LoadPageFromContents(url, cachedContents)
}

func LoadPageFromContents(url, contents string) (*Page, error) {
	pageName, err := GetPageName(url)
	if err != nil {
		return nil, err
	}

	isCharacter := false
	connections := make([]string, 0)
	connectionSet := make(map[string]bool)

	reader := strings.NewReader(contents)
	document, err := goquery.NewDocumentFromReader(reader)
	if err != nil {
		return nil, fmt.Errorf("failed to create parser: %w", err)
	}

	document.Find("div.page-header__categories a").Each(func(_ int, a *goquery.Selection) {
		if value, _ := a.Attr("href"); value == "/wiki/Category:Characters" {
			isCharacter = true
		}
	})

	document.Find("main a").Each(func(_ int, a *goquery.Selection) {
		href, _ := a.Attr("href")
		matches := RegexCharacterHREF.MatchString(href)
		_, visited := connectionSet[href]

		if matches && !visited {
			connections = append(connections, fmt.Sprintf("https://warriors.fandom.com%s", href))
			connectionSet[href] = true
		}
	})

	return &Page{
		URL:         url,
		Name:        pageName,
		IsCharacter: isCharacter,
		Connections: connections,
		Contents:    contents,
	}, nil
}
